### ISSUE TITLE
[FEATURE REQ] ServiceBus: Improve maxConcurrentCalls documentation to prevent production incidents

### ISSUE THREAD

**Issue by gbrehmer** [Azure/azure-sdk-for-java#37192]
https://github.com/Azure/azure-sdk-for-java/blob/8b6f7e4b7bde29f5d0a81eb2415beddc9b682db6/sdk/servicebus/azure-messaging-servicebus/TROUBLESHOOTING.md?plain=1#L26
This important notice should be added to the configuration documentation of max-concurrent-calls and prefetch-count, e.g.: 

https://github.com/Azure/azure-docs-sdk-java/blob/cef4e0be8839c16a71f3d5dc57fd96c7dcbbc41d/docs-ref-autogen/com.azure.messaging.servicebus.ServiceBusClientBuilder.ServiceBusProcessorClientBuilder.yml#L59

If you searching for this exception/behavior its already too late and this parameter has produced an incident in the production environment. We had an incident because we have defined several consumers with prefetch and maxConcurrentCalls, but in our stage system this error does not occur, because we have not all consumers in use there (queues from different azure regions). And it was working before in another production system with more host resources (but this was shared environment, we moved now to AKS and have dedicated but more limited ressources (4 vs 1 CPU)

**Comment by anuchandy** [comment]
Hello @gbrehmer, thanks for reaching out! We’ll improve the Java-Doc here. Just curious, could you let me know the following - 
1.	In AKS you mentioned each Pod has 1 core. How many ProcessorClient instances are you running on one such Pod?
2.	What is the Prefetch and MaxConcurrentCalls set for an instance of ProcessorClient?

**Comment by gbrehmer** [comment]
Sure I forgot that there are several possibilities to achieve the same result:
We are currently not using the programmatic approach. We are using spring-cloud-stream servicebus and for this mainly spring yaml configuration. Because of that javadoc would not automatically help in cases like ours I think, because we made the configuration e.g. on basis of this documentation page: https://learn.microsoft.com/en-us/azure/developer/java/spring-framework/spring-cloud-stream-support?tabs=SpringCloudAzure4x

would you rather recommend a programmatic approach for the configuration?

our use case:
we have in production 4 consumer with in total 16 "concurrent calls". The pod has only one container (spring boot service) with cpu limit `1000m`

```
  cloud:
    stream:
      servicebus:
        bindings:
          serviceBusQueueWestEurope-in-0:
            consumer:
              auto-complete: false
              max-concurrent-calls: 6
              prefetch-count: 5
          serviceBusQueueEastAsia-in-0:
            consumer:
              auto-complete: false
          serviceBusQueueCentralUs-in-0:
            consumer:
              auto-complete: false
              max-concurrent-calls: 6
              prefetch-count: 5
          serviceBusQueueAustraliaEast-in-0:
            consumer:
              auto-complete: false
              max-concurrent-calls: 3
```

**Comment by gbrehmer** [comment]
sure recommended values can be a good starting point, but I would argue, that 2 CPUs and 4 GB RAM are not appropriate values in today's world of microservices and would rather mean a wasteful use of resources.
So in our case we choose another approach by overwriting the JVM CPU count by setting `-XX:ActiveProcessorCount=2`. Our service is not only processing servicebus messages, we also processing a similar amount of rabbitmq messages at the same time and also serving a REST API for client requests. And we didn't saw any resource problems in the past (current avg cpu usage for each pod is < 0.1 CPU), only stuck servicebus processing.
In addition there are multiple other GitHub issues which also describe the same problems with stuck processing and with the unknown `reactor.schedulers.defaultBoundedElasticSize` parameter.

Another idea to prevent such issues would be a check on startup whether `max-concurrent-calls` parameter would fit into the defaultBoundElastic reactor threadpool. Of course there can be other libraries as well that use this threadpool, but in our case this would cause an error already in our dev system.

**Comment by anuchandy** [comment]
1.	Regarding tuning `reactor.schedulers.defaultBoundedElasticSize`, the on-going work related to Prefetch reliability addresses this problem completely so no more interruption/hang due to pool under sizing. For tracking, I’m linking the Prefetch reliability WIP [PR](https://github.com/Azure/azure-sdk-for-java/pull/34854).
2.	Regarding the direct vs indirect (via binders) use of SDK, yes, it’s more of a choice, if you’re already integrated app into the Spring echo system then using binder may make sense. I’ve also seen many users leveraging SDK directly in the Spring app though.
3.	Regarding the core numbers, those are general recommendation coming from a team of experts based on their case studies of AKS Java apps onboarding. I’m not part of that team, so I can’t specifically explain the data points behind the numbers. For my own learning I normally ask users about configuration and saw typical 2 core setup. You can always share your observation and feedback to the java-containerization-team via feedback link at the bottom of the page here https://learn.microsoft.com/en-us/azure/developer/java/containers/overview.

To summarize action scoped to SDK, as mentioned in point 1, removal of the need for users tuning pool-size based on max-concurrent-calls and prefetch-count is an enhancement implemented and is in the pipeline.

**Comment by anuchandy** [comment]
Update on the prefetch reliability work I mentioned in the above comment.

We've shipped new session processor (a.k.a. V2 session processor) as beta. The steps to use -

### Add 7.15.0-beta.5 dependency

```xml
<dependency>
    <groupId>com.azure</groupId>
    <artifactId>azure-messaging-servicebus</artifactId>
    <version>7.15.0-beta.5</version>
</dependency>
```

### Opt-in for new session processor (V2) functionality 

```java
private static final String V2_SESSION_PROCESSOR_CONFIGURATION_KEY = "com.azure.messaging.servicebus.session.processor.asyncReceive.v2";
ServiceBusProcessorClient processor = new ServiceBusClientBuilder()
        .fullyQualifiedNamespace(Config.FULLY_QUALIFIED_NAMESPACE)
        .credential(Config.CREDENTIAL)
        .configuration(new com.azure.core.util.ConfigurationBuilder()
          .putProperty(V2_SESSION_PROCESSOR_CONFIGURATION_KEY, "true") // 'false' by default, so opt-in to enable V2.
          .build())
        .sessionProcessor()
        .topicName(Config.TOPIC)
        .subscriptionName(Config.SUBSCRIPTION)
        .disableAutoComplete()
        .processMessage(processMessage)
        .processError(processError)
        .maxAutoLockRenewDuration(Duration.ofMinutes(5))
        .receiveMode(ServiceBusReceiveMode.PEEK_LOCK)
        .subQueue(SubQueue.NONE)
        .maxConcurrentSessions(2)
        .maxConcurrentCalls(1)
        .prefetchCount(0)
        .sessionIdleTimeout(Duration.ofMinutes(1))
        .buildProcessorClient();
```

**Comment by anuchandy** [comment]
Resolving this. Refer https://learn.microsoft.com/en-us/azure/developer/java/sdk/troubleshooting-messaging-service-bus-overview#upgrade-to-715x for details about version 7.15.x and how to opting-in new session rework.
